src/shader_recompiler/ir/passes/srt.h

```diff
@@ -0,0 +1,126 @@
+// SPDX-FileCopyrightText: Copyright 2024 shadPS4 Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include <vector>
+#include <boost/align/aligned_allocator.hpp>
+#include <boost/align/aligned_delete.hpp>
+#include <boost/container/map.hpp>
+#include <boost/container/set.hpp>
+#include <boost/container/small_vector.hpp>
+
+#include <memory>
+#include "common/alignment.h"
+#include "common/assert.h"
+#include "common/types.h"
+#include "xbyak/xbyak.h"
+
+namespace Shader {
+
+class FlattenedUserDataBuffer {
+public:
+    template <typename T>
+    T ReadUdSharp(u32 sharp_idx) const noexcept {
+        return *reinterpret_cast<const T*>(&buf[sharp_idx]);
+    }
+
+    size_t num_dwords() const {
+        return buf.size();
+    }
+
+    size_t size_bytes() const {
+        return buf.size() * sizeof(u32);
+    }
+
+    u32* data() {
+        return buf.data();
+    }
+
+    const u32* data() const {
+        return buf.data();
+    }
+
+    void resize(size_t new_size_dw) {
+        buf.resize(new_size_dw);
+    }
+
+private:
+    std::vector<u32> buf;
+};
+
+typedef void(__attribute__((sysv_abi)) * PFN_SrtWalker)(const u32* /*user_data*/,
+                                                        u32* /*flat_dst*/);
+
+// Utility for copying a simple relocatable function from a Xbyak code generator to manage memory
+// separately
+class SmallCodeArray {
```

Comment:
This class also raises some questions if its useful or not. If we manage memory externally we can emit to it directly or just let xbyak manage it with a static instance, either way copying can probably be avoided

Reply:
> let xbyak manage it with a static instance
I dont think theres a way to garbage collect old programs if we do this
I thought if u have permutation blow up, then this might become a problem
thats the main reason i did this

But SmallCodeArray is probably superfluous vs having 1 Xbyak::CodeGenerator : 1 Shader::Info.
The only pro to SmallCodeArray is being smaller (forgot exactly why i did it), so probably not worth